### ISSUE TITLE
Eval framework: run-record telemetry contract (Option 1 keystone)

### ISSUE THREAD

**Issue by jstoup111** [jstoup111/ai-conductor#110]
## Summary

Design (and later build) a **run-record telemetry contract** emitted by the harness so that
skill/agent **prompt quality can be evaluated over time** — specifically: *do skill prompts
still produce good artifacts after we edit them?*

This issue is **Option 1 (the keystone): the emitted contract** — not the full eval system.
The eval scoring/dashboard lives in a **separate project** that consumes this contract. Getting
the contract right is the prerequisite for everything downstream; the rest can come later.

> Scope note: this is design/discovery captured from a working session. No code yet.
> Pick up later. The single most important deliverable is the **JSONL run-record schema +
> prompt-hash/version stamping**, because that is the one thing that **cannot be backfilled**.

## Why this exists / the problem

The repo already has *code-quality* evaluation (`agents/evaluator.md`, `/code-review`,
`/prd-audit`). What it does **not** have is a way to regression-test the **skill prompts
themselves** — e.g. after editing `skills/brainstorm/SKILL.md`, did `/brainstorm` get worse at
producing PRDs?

Three things make this hard and shaped the design:
1. **Non-determinism** — same prompt, different output each run. Need sampling + tolerance bands,
   not equality checks.
2. **"Good" is per-skill and partly subjective** — a good PRD ≠ a good plan ≠ a good review.
3. **Grading the grader** — an LLM judge has its own variance/bias; uncalibrated scores look
   rigorous but mean nothing.

### 1. Separate project — but the repo boundary is NOT the real decision
A separate eval repo is fine, **but only if there's a stable, versioned contract at the seam.**
Otherwise it becomes "a separate repo still tightly coupled to harness internals" (`.docs/`
shapes, `conduct-state.json`, `gates/*.json`, retro format) — the worst case, breaking on every
refactor. **Invest in the emitted contract, not the repo split.**

### 2. Reuse the existing event spine
The harness already has the spine for this: `src/conductor/src/ui/events.ts`
(`ConductorEventEmitter`, the `ConductorEvent` union, with `TerminalSubscriber` as one consumer).
The eval pipeline should be **just another subscriber/sink** of that event stream, plus a reader
of the artifacts those events point to.

- **Harness side (small, this issue):** add a telemetry sink subscriber that writes a durable,
  versioned **run-record**: append-only JSONL events + a manifest of artifact paths/hashes +
  gate verdicts + final outcome.
- **Eval project (separate repo, later):** consumes run-records from many real executions,
  scores them, stores time-series, detects regressions, dashboards.

### 3. Live (online) AND lab (replay) — complementary, not either/or
- **Live tracking** (real runs) = the **data + observability** layer. Real input distribution,
  free data, captures kickbacks/halts/rework. **But it cannot do attribution** — every run has
  different inputs, so a quality dip could be your prompt edit, a harder task, or model drift.
- **Lab / replay** (golden cases) = the **causation** layer. Re-run the same case before/after a
  prompt edit.
- **They connect:** snapshot real runs from the live stream → replay them as golden cases. Live
  telemetry is the corpus generator for lab evals → solves the "fixture realism" problem; stop
  hand-authoring fixtures.

### 4. The non-backfillable must-have: stamp every event
Every emitted event must carry **(a) harness version and (b) content-hash of the skill prompt
that actually ran** (and ideally model + token/duration). This is the **join key** that lets the
eval project correlate "quality moved" with "this `SKILL.md` changed." Cheap, additive, and
**impossible to backfill** — do this first, before any scoring exists.

### 5. Retro is a signal source, not the thing
`/retro` is the **qualitative, single-run, narrative** cousin (`.docs/retros/`). It already
captures harness-vs-product friction, kickbacks, what failed — but it's per-feature, not
aggregated, can't detect regressions. The eval project should **ingest retros (and audits) as
one telemetry stream**; eval is the *quantitative aggregate across many runs* that no single
retro can be.

## What "Option 1" concretely is (the scoped work for pickup)

1. **Read the current surface** before designing:
   - `src/conductor/src/ui/events.ts` — `ConductorEvent` union, emitter, subscriber pattern.
   - `src/conductor/src/ui/subscriber.ts` (`TerminalSubscriber`) — model the sink on this.
   - How the daemon persists today: `.pipeline/conduct-state.json` and `.pipeline/gates/*.json`
     (see `src/conductor/src/engine/state.ts`, `gates.ts`, `gate-verdicts.ts`).
   - `.docs/` artifact tree (specs/stories/plans/architecture/audits/retros) — what each run
     already leaves on disk.
2. **Spec the run-record JSONL schema** (the contract): event entries + artifact manifest
   (paths + content hashes) + gate verdicts + final outcome (shipped / halted / N kickbacks) +
   metadata (harness version, per-step skill-prompt content-hash, model, tokens, duration).
3. **Add the telemetry sink** as a new `ConductorEventEmitter` subscriber writing the run-record
   to a known location (and/or shippable to the eval project).
4. **Add prompt-hash + version stamping** to emitted events (the non-backfillable bit).
5. Versioned + documented so the external eval repo can depend on it as a stable API.

## Inventory worth doing alongside (was "Option 2")
Map what every real run **already** leaves on disk (events, `conduct-state.json`, `gates/*.json`,
retros, audits) to see how much of "live tracking" is just *collect + ship what exists* vs. new
instrumentation. Likely most of it already exists; the gap is a durable append-only event log +
stamping + shipping.

## Downstream / later (NOT this issue, captured so it isn't lost)
- **Grading = hybrid:** deterministic gate first (reuse `gates.ts` / `artifacts.ts`
  `CUSTOM_COMPLETION_PREDICATES` — these already encode "what good looks like"; structural fail =
  hard zero), then **LLM-judge** (new `agents/eval-judge.md`, fresh context, 0–1 + rationale,
  median-of-N) only on what passes structurally.
- **Highest-signal first eval = seeded-defect** for evaluator/`/code-review`: plant a known bug
  in a diff, assert it's caught (recall/precision, no judge variance). Ready-made labeled corpus:
  the "critical bug in 3/3 Phase 9 features" negative-path gap.
- **Phase 0 de-risk experiments (do before building scoring):** (a) run one skill on one fixture
  ~10× and measure score stddev → sets the regression threshold + samples-per-case; (b) judge
  calibration: hand-label ~10 artifacts, measure judge agreement.
- **Cost/cadence:** 22 skills × N samples × Opus judge is expensive → nightly or on-skill-PR,
  never per-commit.
- **Headless invocation constraint:** interactive skills (`/brainstorm`, `/engineer`) ask
  clarifying questions; reuse the daemon's non-interactive/auto path or start with
  non-interactive skills.

## Acceptance for THIS issue (when picked up)
- [ ] A written spec of the run-record JSONL schema + artifact manifest (the contract), versioned.
- [ ] Telemetry sink subscriber emitting it, modeled on `TerminalSubscriber`.
- [ ] Every event stamped with harness version + per-step skill-prompt content-hash (+ model,
      tokens, duration).
- [ ] Docs describing the contract as the stable seam the external eval project consumes.

---

## Operator scope extension — evaluation-driven skill improvement (2026-09-11)

This extends the original telemetry-only scope above to include implementation of a skill self-improvement workflow. The earlier “downstream / later (NOT this issue)” boundary is superseded for this capability; the telemetry contract remains a foundation. Repo placement and integration design remain DECIDE decisions.

### Observed

The operator requested: “i'd like to add https://github.com/microsoft/skillopt implementation (or similar tool) to self improve skills to get better outputs.”

Verified from [Microsoft's SkillOpt README](https://github.com/microsoft/SkillOpt#overview): SkillOpt optimizes natural-language skill documents using scored execution trajectories and bounded edits, without changing target-model weights. Its default paper-style path accepts candidates only when held-out validation improves; the resulting artifact is `best_skill.md`. These are documented upstream capabilities, not verified ai-conductor integration results.

Related intake: #1612 tracks durable, per-skill evaluation corpora and replay, including an operator extension beyond rubric skills. Its latest comment calls for revalidation against the changed review architecture. #807 tracks flow execution correctness; this extension concerns the quality of outputs produced by skills.

### Impact

The operator wants skill changes to produce demonstrably better outputs, with an automated improvement loop and evidence that separates improvement from regressions. Telemetry alone does not deliver that outcome; no ai-conductor quality gain from SkillOpt is claimed yet.

### Desired outcome

- An operator can run an end-to-end cycle for a selected existing skill: evaluate the baseline, generate candidate skill revisions from execution feedback, evaluate the candidates, and obtain a reviewable proposed revision with supporting results.
- Baseline and candidates are compared on the same representative tasks with recorded skill versions, model/provider configuration, grading criteria, and repeat-run variability. Evaluation measures task/artifact quality, not only whether the skill activates.
- Optimization examples are separated from held-out evaluation cases. A candidate is eligible for promotion only when it meets a declared improvement threshold and regression constraints; failed or inconclusive candidates leave the active skill unchanged.
- At least one existing ai-conductor skill receives an implementation-backed pilot with recorded before/after results. A negative result is reported honestly rather than labeled an improvement.
- Proposed changes preserve skill contracts and required gates, remain reviewable before adoption, and can be rolled back to the prior version.
- Runs respect explicit cost/iteration limits and report quality, failures, and resource usage. Real-model evaluation/optimization is opt-in; the default automated test suite and CI use faithful fakes at third-party boundaries, with real-service tests restricted to the named opt-in smoke lane.
- The workflow respects the selected host/provider and per-project, per-skill corpus boundaries. Missing compatible execution support is reported explicitly.

### Hypotheses

Candidates for evaluation; the operator requested a broader search beyond SkillOpt on 2026-09-11. DECIDE must validate fit before selecting an implementation:

- Integrate Microsoft SkillOpt, or a comparable existing optimizer if it better fits ai-conductor's skill contracts and provider execution. Compare candidate editing, held-out validation, replay support, cost controls, and maintainability before selecting.
- Reuse the run-record contract and per-skill corpus work in #1612 for inputs and attribution. Revalidate the available corpus and replayable inputs against current code; historical issue descriptions are not proof of current readiness.
- For harness telemetry, extend the existing ConductorEvent spine and its reader path; do not introduce a competing observation channel. Candidate skill documents and evaluation datasets are durable artifacts, distinct from run events.
- A bounded offline pilot can establish whether the optimizer improves a selected skill before expanding coverage or scheduling recurring optimization.

Claim verification: CLEAR for intake. Upstream capabilities are verified from the linked documentation; integration feasibility and quality gains remain open evaluation questions, not assumptions used to approve a design.

### Hypotheses — alternatives beyond SkillOpt (2026-09-11)

The operator requested other solutions beyond SkillOpt. Capabilities below are verified from linked upstream documentation; fit assessments are engineering inferences (~80% confidence), not results from an ai-conductor integration or benchmark.

| Candidate | Documented capability | Potential fit and integration question |
| --- | --- | --- |
| [GEPA `optimize_anything`](https://gepa-ai.github.io/gepa/api/optimize_anything/optimize_anything/) ([repo](https://github.com/gepa-ai/gepa)) | Optimizes a text candidate or named text components through a supplied evaluator; exposes dataset, validation, and test-set inputs. [Agent-skill guide](https://github.com/gepa-ai/gepa/blob/main/docs/docs/guides/agent-skill.md) documents its coding-agent workflow. | Strong direct candidate: treat a skill as the candidate and run existing ai-conductor execution behind the evaluator. Validate preservation of frontmatter/contracts and the cost of realistic replay. |
| [Anthropic skill-creator](https://github.com/anthropics/skills/blob/main/skills/skill-creator/SKILL.md) | Iterates on skill outputs using test prompts, old-skill/no-skill baselines, quantitative grading, human review, and optional blind comparison. Has a separate automated description/trigger optimization loop. | Useful direct skill-authoring baseline and source of evaluation/review practices. Its output-quality loop is agent/human-driven; the automated description loop alone does not establish better artifact quality. Provider-specific execution needs adaptation. |
| [DSPy MIPROv2](https://dspy.ai/api/optimizers/MIPROv2/) | Jointly optimizes instructions and few-shot examples using candidate generation and Bayesian optimization; can optimize instructions alone. | Useful algorithmic comparison if skills benefit from example selection. Determine whether wrapping current skill execution in a DSPy program is worthwhile; do not assume a drop-in SKILL.md editor. DSPy's GEPA integration is another entry point to GEPA, not an independent optimizer baseline. |
| [Microsoft Agent Lightning APO](https://github.com/microsoft/agent-lightning/blob/main/examples/apo/README.md) | Provides an iterative prompt-template optimization example with beam-search parameters and a training loop. | Candidate if broader agent-training orchestration is needed. Verify adapter and deployment overhead; the example optimizes a prompt template, not a complete ai-conductor skill package. |
| [ACE — Agentic Context Engineering](https://github.com/ace-agent/ace) | Evolves contextual playbooks through generation, reflection, and curation. | Candidate for accumulating reusable lessons from execution history. Validate how playbooks map to governed skill revisions and how independent evaluation gates promotion. |
| [Promptfoo](https://www.promptfoo.dev/docs/configuration/guide/) | Evaluates prompts/providers against test cases and assertions; supports custom JavaScript/Python scoring and integrations. | Candidate for the shared comparison/regression layer alongside an optimizer. The documented evaluation workflow is not by itself an autonomous skill-rewriting implementation. |

Operator direction (2026-09-11): retain these alternatives as a reference list. The operator will run the POC and choose what to pursue; no candidate or evaluation order is selected by this issue update.

Clarification to the older skill-creator discussion on this issue: the currently inspected upstream skill includes output-quality evaluation and revision as well as a separate description-trigger loop. Describing the entire skill-creator workflow as activation testing would be too narrow; the earlier statement about `run_eval.py` specifically should not be generalized to the whole tool.

**Comment by jstoup111** [comment]
## Prior art worth evaluating before building: Anthropic's `skill-creator`

Surfaced while investigating an unrelated hotfix. `skill-creator` is an Anthropic-authored plugin in the official marketplace (`anthropics/claude-plugins-official`) that already implements part of what this issue's downstream consumer would need. It is **not** a substitute for the keystone here — see the seam below — but it is real, working prior art rather than a design sketch.

**Location (already cloned locally, not installed):** `~/.claude/plugins/marketplaces/claude-plugins-official/plugins/skill-creator/`. 21 files: a 485-line SKILL.md plus 9 Python scripts.

### What it actually does — verified from the scripts, not the README

- **`run_eval.py`** (310 lines) — *"Tests whether a skill's description causes Claude to trigger (read the skill) for a set of queries."* This is **activation testing**, not artifact-quality testing.
- **`aggregate_benchmark.py`** (401 lines) — reads per-run `grading.json` files and produces *"run_summary with mean, stddev, min, max for each metric"* plus *"delta between `with_skill` and `without_skill` configurations."*
- **`improve_description.py`** (247 lines) — rewrites a description from eval results by *"calling `claude -p` as a subprocess."*
- Also `quick_validate.py`, `package_skill.py`, `run_loop.py`, `generate_report.py`.

### How it relates to this issue's scope

**It answers a different question.** This issue asks *"after editing `skills/<x>/SKILL.md`, did the skill get worse at producing artifacts?"* — measured from **run-records emitted by real harness runs**, which is the part explicitly called out here as impossible to backfill. `run_eval.py` asks *"does this description get the skill read at all?"*, evaluated in isolation against synthetic queries. Neither subsumes the other, and nothing in `skill-creator` emits a run-record from a production run. **The keystone deliverable in this issue is unaffected.**

**Two pieces are directly reusable as design input, though:**

1. `aggregate_benchmark.py`'s mean/stddev/min/max over N runs plus a `with_skill` vs `without_skill` **control condition** is a concrete instance of design problem #1 in this issue (*"non-determinism — need sampling + tolerance bands, not equality checks"*). Worth reading before designing our own tolerance bands.
2. The triggering axis is a **real gap we have today, independent of this issue**. This repo ships 33 skills in `skills/` plus 4 in `.agents/skills/`, all competing for activation on their frontmatter `description`, and none of that is tested. A misfiring description silently degrades a whole SDLC phase and would never show up in artifact-quality scoring.

### Adoption caveat — flagging early because it is easy to miss

`improve_description.py` shells out to **`claude -p`**. That collides with two standing rules here: the repository's test-isolation policy (third-party/LLM calls are smoke-only, excluded from the default suite and CI), and the engineer loop's explicit *"zero `claude -p` / authoring subprocess"* boundary. Any adoption would need those scripts confined to an opt-in smoke lane, or the `claude -p` call replaced with the harness's own provider abstraction. Reusing the *approach* is cheap; vendoring the scripts as-is is not.

### On picking it by popularity — you cannot

All 40 plugins are directories inside one repo (33,283 stars, 3,758 forks as of 2026-08-09). The marketplace manifest carries only `name`/`description`/`author`/`category`/`homepage` — no install counts, no per-plugin stars. Author is Anthropic, so there is no third-party trust question, but there is also no popularity signal to rank on. Fit is the only criterion available.

### Suggested next step

Read `skill-creator`'s SKILL.md and `aggregate_benchmark.py` as input to this issue's contract design, and treat **skill-description triggering coverage** as a candidate sibling ticket rather than folding it into this one — it needs no run-record contract and could land independently.

**Comment by jstoup111** [comment]
### Correction to the comment above: `claude plugin eval` is already built in

My previous comment framed `skill-creator`'s Python scripts as the notable prior art. That undersold what is available, and in one respect pointed at the wrong thing.

Claude Code ships a first-class eval command — no plugin required:

    claude plugin eval [options] [target]
      Run eval cases (evals/**/case.yaml or evals/**/prompt.md + graders/*.md)
      against a plugin and report scored results. Target is a path, a plugin name,
      or a `plugin@marketplace` id — installed and skills-dir plugins both resolve
      (and add a no-plugin baseline arm)

Three things follow for this issue:

1. **The baseline arm is native.** The `with_skill` vs `without_skill` control condition I cited from `aggregate_benchmark.py` is not something we would need to reimplement or vendor — `claude plugin eval` adds a no-plugin baseline arm itself. Design problem #1 in this issue (non-determinism, sampling, tolerance bands rather than equality checks) has a supported starting point.

2. **There is a declarative case format to evaluate before inventing one:** `evals/**/case.yaml`, or `evals/**/prompt.md` paired with `graders/*.md`. Whatever run-record schema this issue lands should be checked against that shape first — if the two can be made compatible, the downstream scoring project gets a working runner for free instead of building one.

3. **The `claude -p` adoption caveat I raised is narrower than I implied.** It applies to `improve_description.py` specifically, not to evaluating skills in general. The built-in path does not require vendoring that script, so the collision with this repo's test-isolation policy and the engineer loop's zero-`claude -p` boundary is avoidable rather than something to design around.

None of this changes the keystone. `claude plugin eval` scores a skill or plugin against authored eval cases; it does not emit a run-record from a real harness run, which is the thing this issue identifies as impossible to backfill. **The contract is still the prerequisite.** What changes is the build-versus-adopt calculus for the consumer that sits downstream of it.

Also relevant to the sibling ticket I suggested (skill-description triggering coverage): `claude plugin eval` resolves `skills-dir` plugins as targets, so the 33 skills in `skills/` and the 4 in `.agents/skills/` may be testable for triggering without any new harness at all. That makes the sibling ticket smaller than I estimated.
